validation/phase_test.go

```diff
@@ -0,0 +1,602 @@
+					},
+				},
+			},
+			owner: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"metadata": map[string]interface{}{
+						"namespace": "default",
+					},
+				},
+			},
+			mockSetup: func(objValidator *mockObjectValidator) {
+				objValidator.On("Validate", mock.Anything, mock.Anything, mock.Anything).Return(nil)
+			},
+			expectError: false,
+		},
+		{
+			name: "invalid phase name",
+			phase: types.Phase{
+				Name: "Invalid_Phase_Name",
+				Objects: []unstructured.Unstructured{
+					{
+						Object: map[string]interface{}{
+							"apiVersion": "v1",
+							"kind":       "ConfigMap",
+							"metadata": map[string]interface{}{
+								"name":      "test1",
+								"namespace": "default",
+							},
+						},
+					},
+				},
+			},
+			owner: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"metadata": map[string]interface{}{
+						"namespace": "default",
+					},
+				},
+			},
+			mockSetup: func(objValidator *mockObjectValidator) {
+				objValidator.On("Validate", mock.Anything, mock.Anything, mock.Anything).Return(nil)
+			},
+			expectError:              true,
+			expectPhaseValidationErr: true,
+		},
+		{
+			name: "object validation error from mock",
+			phase: types.Phase{
+				Name: "valid-phase",
+				Objects: []unstructured.Unstructured{
+					{
+						Object: map[string]interface{}{
+							"apiVersion": "v1",
+							"kind":       "ConfigMap",
+							"metadata": map[string]interface{}{
+								"name":      "test1",
+								"namespace": "default",
+							},
+						},
+					},
+				},
+			},
+			owner: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"metadata": map[string]interface{}{
+						"namespace": "default",
+					},
+				},
+			},
+			mockSetup: func(objValidator *mockObjectValidator) {
+				objErr := &ObjectValidationError{
+					ObjectRef: testObjRef,
+					Errors:    []error{errTest},
+				}
+				objValidator.On("Validate", mock.Anything, mock.Anything, mock.Anything).Return(objErr)
+			},
+			expectError:              true,
+			expectPhaseValidationErr: true,
+			useRealValidator:         false,
+		},
+		{
+			name: "unknown error during object validation",
+			phase: types.Phase{
+				Name: "valid-phase",
+				Objects: []unstructured.Unstructured{
+					{
+						Object: map[string]interface{}{
+							"apiVersion": "v1",
+							"kind":       "ConfigMap",
+							"metadata": map[string]interface{}{
+								"name":      "test1",
+								"namespace": "default",
+							},
+						},
+					},
+				},
+			},
+			owner: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"metadata": map[string]interface{}{
+						"namespace": "default",
+					},
+				},
+			},
+			mockSetup: func(objValidator *mockObjectValidator) {
+				objValidator.On("Validate", mock.Anything, mock.Anything, mock.Anything).Return(
+					errors.New("unknown error"))
+			},
+			expectError:              true,
+			expectPhaseValidationErr: false,
```

Comment:
This feels a bit surprising to me. I guess the assumption here is that we know the universe of errors we would expect for phase validation and everything else is an error unrelated to phase validation?

Even so, it seems like an unknown error encountered during phase validation should still bubble up as a phase validation error, at least in the form of `UnknownObjectError{err: errors.New("unknown error")}`.

But take this with a grain of salt, since I still don't fully understand the distinction.